File: src/metrics.py
"""Risk and performance metrics. All pure, type-hinted, no I/O."""

from __future__ import annotations

import warnings
from typing import Iterable

import numpy as np
import pandas as pd

ArrayLike = pd.Series | np.ndarray | Iterable[float]


def _as_array(x: ArrayLike) -> np.ndarray:
    if isinstance(x, pd.Series):
        return x.to_numpy(dtype=float)
    return np.asarray(list(x), dtype=float) if not isinstance(x, np.ndarray) else x.astype(float)


def cumulative_returns(daily_returns: ArrayLike) -> pd.Series:
    """Convert a series of simple daily returns into a cumulative-return curve.

    cum[i] = (1 + r_0)(1 + r_1)...(1 + r_i) - 1
    """
    arr = _as_array(daily_returns)
    if arr.size == 0:
        return pd.Series(dtype=float)
    curve = np.cumprod(1.0 + arr) - 1.0
    if isinstance(daily_returns, pd.Series):
        return pd.Series(curve, index=daily_returns.index)
    return pd.Series(curve)


def portfolio_returns(weights: ArrayLike, asset_returns: pd.DataFrame) -> pd.Series:
    """Daily returns of a fixed-weight portfolio."""
    w = _as_array(weights)
    if w.shape[0] != asset_returns.shape[1]:
        raise ValueError(
            f"weights length {w.shape[0]} != number of assets {asset_returns.shape[1]}"
        )
    if not np.isclose(w.sum(), 1.0, atol=1e-6):
        warnings.warn(f"weights sum to {w.sum():.6f}, not 1.0", stacklevel=2)
    return asset_returns.dot(w)


def annualized_return(daily_returns: ArrayLike, factor: int = 365) -> float:
    """Mean daily return * factor."""
    arr = _as_array(daily_returns)
    if arr.size == 0:
        return float("nan")
    return float(np.mean(arr) * factor)


def annualized_volatility(daily_returns: ArrayLike, factor: int = 365) -> float:
    """Std of daily returns * sqrt(factor)."""
    arr = _as_array(daily_returns)
    if arr.size < 2:
        return float("nan")
    return float(np.std(arr, ddof=1) * np.sqrt(factor))


def sharpe_ratio(
    daily_returns: ArrayLike, risk_free_rate: float = 0.0, factor: int = 365
) -> float:
    """(annual return - rf) / annual vol. NaN when vol is 0."""
    vol = annualized_volatility(daily_returns, factor)
    if not np.isfinite(vol) or vol == 0.0:
        return float("nan")
    return (annualized_return(daily_returns, factor) - risk_free_rate) / vol


def sortino_ratio(
    daily_returns: ArrayLike, risk_free_rate: float = 0.0, factor: int = 365
) -> float:
    """Like Sharpe but uses downside deviation (returns below 0)."""
    arr = _as_array(daily_returns)
    if arr.size < 2:
        return float("nan")
    downside = arr[arr < 0]
    if downside.size == 0:
        return float("nan")
    downside_vol = float(np.std(downside, ddof=1) * np.sqrt(factor))
    if not np.isfinite(downside_vol) or downside_vol == 0.0:
        return float("nan")
    ann_ret = annualized_return(arr, factor)
    return (ann_ret - risk_free_rate) / downside_vol


def max_drawdown(daily_returns: ArrayLike) -> float:
    """Worst peak-to-trough drop on the cumulative wealth curve. Returns a non-positive number."""
    arr = _as_array(daily_returns)
    if arr.size == 0:
        return float("nan")
    wealth = np.cumprod(1.0 + arr)
    peak = np.maximum.accumulate(wealth)
    drawdown = wealth / peak - 1.0
    return float(drawdown.min())


def value_at_risk(daily_returns: ArrayLike, level: float = 0.95) -> float:
    """Historical VaR. Returns a non-negative number representing the loss magnitude.

    e.g. 0.04 means 4% worst-case daily loss at the (1-level) tail.
    """
    arr = _as_array(daily_returns)
    if arr.size == 0:
        return float("nan")
    quantile = float(np.quantile(arr, 1.0 - level))
    return float(-quantile) if quantile < 0 else 0.0


def conditional_value_at_risk(daily_returns: ArrayLike, level: float = 0.95) -> float:
    """Mean loss in the worst (1-level) tail. Non-negative magnitude."""
    arr = _as_array(daily_returns)
    if arr.size == 0:
        return float("nan")
    cutoff = np.quantile(arr, 1.0 - level)
    tail = arr[arr <= cutoff]
    if tail.size == 0:
        return float("nan")
    mean_tail = float(np.mean(tail))
    return float(-mean_tail) if mean_tail < 0 else 0.0


def portfolio_objective(
    weights: ArrayLike,
    asset_returns: pd.DataFrame,
    mu: np.ndarray,
    Sigma: np.ndarray,
    mode: str,
    risk_aversion: float = 1.0,
    risk_free_rate: float = 0.0,
    factor: int = 365,
) -> float:
    """Compute the objective value to minimize for a given mode.

    All modes return a *minimization* target — Sharpe-like maxes are negated internally.
    """
    w = _as_array(weights)
    daily = portfolio_returns(w, asset_returns)
    ann_ret = float(w @ mu)
    ann_var = float(w @ Sigma @ w)
    ann_vol = float(np.sqrt(max(ann_var, 0.0)))
    if mode == "maximize_sharpe":
        if ann_vol == 0.0:
            return float("inf")
        return -((ann_ret - risk_free_rate) / ann_vol)
    if mode == "minimize_volatility":
        return ann_vol
    if mode == "maximize_return":
        return -ann_ret
    if mode == "maximize_return_minus_risk":
        return -(ann_ret - risk_aversion * ann_var)
    if mode == "minimize_cvar":
        return conditional_value_at_risk(daily)
    if mode == "constrained_sharpe":
        if ann_vol == 0.0:
            return float("inf")
        return -((ann_ret - risk_free_rate) / ann_vol)
    raise ValueError(f"unknown objective mode: {mode}")


def summarize_weights(
    weights: ArrayLike,
    asset_returns: pd.DataFrame,
    mu: np.ndarray,
    Sigma: np.ndarray,
    risk_free_rate: float = 0.0,
    risk_aversion: float = 1.0,
    factor: int = 365,
) -> dict:
    """All metrics for one weight vector. Used by every optimizer + benchmark + quantum result."""
    w = _as_array(weights)
    daily = portfolio_returns(w, asset_returns)
    ann_ret = annualized_return(daily, factor)
    ann_vol = annualized_volatility(daily, factor)
    sharpe = sharpe_ratio(daily, risk_free_rate, factor)
    sortino = sortino_ratio(daily, risk_free_rate, factor)
    mdd = max_drawdown(daily)
    var = value_at_risk(daily)
    cvar = conditional_value_at_risk(daily)
    cum = cumulative_returns(daily)
    final_cum = float(cum.iloc[-1]) if len(cum) else float("nan")
    obj_mv = float(risk_aversion * (w @ Sigma @ w) - (w @ mu))
    return {
        "annual_return": ann_ret,
        "annual_volatility": ann_vol,
        "sharpe_ratio": sharpe,
        "sortino_ratio": sortino,
        "max_drawdown": mdd,
        "var_95": var,
        "cvar_95": cvar,
        "final_cumulative_return": final_cum,
        "objective_mv": obj_mv,
    }

File: julia/src/Metrics.jl
module Metrics

using Statistics
using DataFrames

export cumulative_returns, portfolio_returns, annualized_return,
       annualized_volatility, sharpe_ratio, sortino_ratio,
       max_drawdown, value_at_risk, conditional_value_at_risk,
       portfolio_objective, summarize_weights

function cumulative_returns(r::AbstractVector{<:Real})
    isempty(r) && return Float64[]
    return cumprod(1.0 .+ r) .- 1.0
end

function portfolio_returns(w::AbstractVector{<:Real}, R::DataFrame)
    n_assets = ncol(R)
    length(w) == n_assets ||
        throw(ArgumentError("weights length $(length(w)) != number of assets $n_assets"))
    return Matrix(R) * collect(Float64.(w))
end

function annualized_return(r::AbstractVector{<:Real}; factor::Int=365)
    isempty(r) && return NaN
    return mean(r) * factor
end

function annualized_volatility(r::AbstractVector{<:Real}; factor::Int=365)
    length(r) < 2 && return NaN
    return std(r; corrected=true) * sqrt(factor)
end

# Julia's std on a constant series returns ~1e-18 (rounding), not exact 0
# like NumPy. Treat anything below 1e-12 as "effectively zero" so the
# zero-vol branch fires consistently.
const VOL_EPSILON = 1e-12

function sharpe_ratio(r::AbstractVector{<:Real}; risk_free_rate::Real=0.0, factor::Int=365)
    vol = annualized_volatility(r; factor=factor)
    (isnan(vol) || abs(vol) < VOL_EPSILON) && return NaN
    return (annualized_return(r; factor=factor) - risk_free_rate) / vol
end

function sortino_ratio(r::AbstractVector{<:Real}; risk_free_rate::Real=0.0, factor::Int=365)
    length(r) < 2 && return NaN
    downside = filter(<(0.0), r)
    isempty(downside) && return NaN
    dvol = std(downside; corrected=true) * sqrt(factor)
    (isnan(dvol) || abs(dvol) < VOL_EPSILON) && return NaN
    return (annualized_return(r; factor=factor) - risk_free_rate) / dvol
end

function max_drawdown(r::AbstractVector{<:Real})
    isempty(r) && return NaN
    wealth = cumprod(1.0 .+ r)
    peak = accumulate(max, wealth)
    return minimum(wealth ./ peak .- 1.0)
end

# VaR / CVaR are reported as non-negative loss magnitudes — see metrics.py:100-122
function value_at_risk(r::AbstractVector{<:Real}; level::Real=0.95)
    isempty(r) && return NaN
    q = quantile(collect(Float64, r), 1.0 - level)
    return q < 0 ? -q : 0.0
end

function conditional_value_at_risk(r::AbstractVector{<:Real}; level::Real=0.95)
    isempty(r) && return NaN
    cutoff = quantile(collect(Float64, r), 1.0 - level)
    tail = filter(<=(cutoff), r)
    isempty(tail) && return NaN
    m = mean(tail)
    return m < 0 ? -m : 0.0
end

function portfolio_objective(w::AbstractVector{<:Real}, R::DataFrame,
                             μ::AbstractVector{<:Real}, Σ::AbstractMatrix{<:Real};
                             mode::AbstractString, risk_aversion::Real=1.0,
                             risk_free_rate::Real=0.0, factor::Int=365)
    daily = portfolio_returns(w, R)
    ann_ret = dot(w, μ)
    ann_var = dot(w, Σ * w)
    ann_vol = sqrt(max(ann_var, 0.0))
    if mode == "maximize_sharpe" || mode == "constrained_sharpe"
        ann_vol == 0.0 && return Inf
        return -(ann_ret - risk_free_rate) / ann_vol
    elseif mode == "minimize_volatility"
        return ann_vol
    elseif mode == "maximize_return"
        return -ann_ret
    elseif mode == "maximize_return_minus_risk"
        return -(ann_ret - risk_aversion * ann_var)
    elseif mode == "minimize_cvar"
        return conditional_value_at_risk(daily)
    else
        throw(ArgumentError("unknown objective mode: $mode"))
    end
end

# helper since Julia stdlib's `dot` is in LinearAlgebra (imported here for self-containment)
using LinearAlgebra: dot

"""
    summarize_weights(w, R, μ, Σ; risk_free_rate, risk_aversion, factor)

All metrics for one weight vector. Field names match
[src/metrics.py:185-195](src/metrics.py:185) so downstream CSV columns line up.
"""
function summarize_weights(w::AbstractVector{<:Real}, R::DataFrame,
                           μ::AbstractVector{<:Real}, Σ::AbstractMatrix{<:Real};
                           risk_free_rate::Real=0.0, risk_aversion::Real=1.0,
                           factor::Int=365)
    daily = portfolio_returns(w, R)
    cum = cumulative_returns(daily)
    return (
        annual_return = annualized_return(daily; factor=factor),
        annual_volatility = annualized_volatility(daily; factor=factor),
        sharpe_ratio = sharpe_ratio(daily; risk_free_rate=risk_free_rate, factor=factor),
        sortino_ratio = sortino_ratio(daily; risk_free_rate=risk_free_rate, factor=factor),
        max_drawdown = max_drawdown(daily),
        var_95 = value_at_risk(daily),
        cvar_95 = conditional_value_at_risk(daily),
        final_cumulative_return = isempty(cum) ? NaN : last(cum),
        objective_mv = risk_aversion * dot(w, Σ * w) - dot(w, μ),
    )
end

end # module
